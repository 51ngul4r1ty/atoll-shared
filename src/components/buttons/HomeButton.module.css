.button {
    align-items: center;
    background-color: var(--inverted-page-color, black);
    cursor: pointer;
    display: inline-flex;
    padding: 0.5rem 0.85rem 0.5rem 0.7rem;
    user-select: none;
}

.button:focus,
.buttonFocus {
    background-color: var(--inverted-focus-page-color, #444);
    outline: 0;
}

.button:hover,
.buttonHover {
    box-shadow: 0 0 0 rgba(0, 0, 0, 0) inset,
        0 0 var(--normal-hover-shadow-size, 0.6rem) var(--normal-hover-shadow, rgba(0, 0, 0, 0.4));
}

.button:active,
.buttonActive {
    background-color: var(--inverted-active-page-color, #555);
}

.buttonCaption {
    color: var(--inverted-pen-color, white);
    font-size: var(--large-font-size, 1.3rem);
    font-weight: var(--primary-bold-font-weight, 700);
}

.buttonIcon + .buttonCaption {
    padding-left: 0.5rem;
}

.buttonIcon {
    display: inline-flex;
}

.buttonIcon > * {
    height: 1.9rem;
    width: 1.9rem;
}

.buttonIcon * {
    font-size: 1px;
}
