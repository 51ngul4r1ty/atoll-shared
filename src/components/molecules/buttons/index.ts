export { AddButton } from "./AddButton";
export { ArchiveButton } from "./ArchiveButton";
export { CancelButton } from "./CancelButton";
export { DoneButton } from "./DoneButton";
export { EditButton } from "./EditButton";
export { FrameCloseButton } from "./FrameCloseButton";
export { HomeButton } from "./HomeButton";
export { RefreshButton } from "./RefreshButton";
export { RemoveButton } from "./RemoveButton";
export { ResetButton } from "./ResetButton";
export { SaveButton } from "./SaveButton";
export * from "./SimpleButton";
