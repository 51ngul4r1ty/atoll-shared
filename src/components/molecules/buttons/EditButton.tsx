// externals
import * as React from "react";

// atoms
import { SimpleButton } from "../../atoms/buttons/SimpleButton";

// style
import css from "./EditButton.module.css";
import { EditIcon } from "../../atoms/icons/EditIcon";
import { EditCancelIcon } from "../../atoms/icons/EditCancelIcon";

export enum EditMode {
    VIEW,
    EDIT
}

export interface EditButtonAttributeProps {
    mode: EditMode;
}

export interface EditButtonEventProps {
    onClick: { () };
}

export type EditButtonProps = EditButtonAttributeProps & EditButtonEventProps;

export const EditButton: React.FC<EditButtonProps> = (props) => {
    const icon = props.mode === EditMode.VIEW ? <EditIcon /> : <EditCancelIcon />;
    const text = props.mode === EditMode.VIEW ? "Edit" : "View";
    return (
        <SimpleButton iconOnLeft icon={icon} onClick={props.onClick}>
            {text}
        </SimpleButton>
    );
};
