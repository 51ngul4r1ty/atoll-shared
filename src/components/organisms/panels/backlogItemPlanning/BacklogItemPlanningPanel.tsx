// externals
import * as React from "react";
import { useState } from "react";
import { withTranslation } from "react-i18next";

// style
import css from "./BacklogItemPlanningPanel.module.css";

// components
import {
    BacklogItemCard,
    BacklogItemTypeEnum,
    buildBacklogItemKey,
    buildBacklogItemPlanningItemKey,
    buildDividerKey,
    ItemMenuEventHandlers
} from "../../../molecules/cards/BacklogItemCard";
import { SimpleDivider } from "../../../atoms/dividers/SimpleDivider";

// consts/enums
import { EditMode } from "../../../molecules/buttons/EditButton";
import { buildClassName } from "../../../../utils/classNameBuilder";
import { useDispatch } from "react-redux";
import { SaveableBacklogItem } from "../../../../reducers/backlogItems/backlogItemsReducerTypes";
import { Source } from "../../../../reducers/types";

// actions
import { apiDeleteBacklogItem } from "../../../../actions/apiBacklogItems";
import {
    backlogItemDetailClicked,
    editBacklogItem,
    selectProductBacklogItem,
    unselectProductBacklogItem
} from "../../../../actions/backlogItemActions";

// utils
import * as logger from "../../../../utils/logger";
import { useRecursiveTimeout } from "../../../common/setTimeoutHook";
import { BacklogItemPlanningItem } from "../../combo/BacklogItemPlanningItem";
import { buildBacklogDisplayId } from "../../../../utils/backlogItemHelper";

// consts/enums
import * as loggingTags from "../../../../constants/loggingTags";
import { BacklogItemPlanningPanelProps, CardPosition } from "./backlogItemPlanningPanelTypes";
import {
    atBottomOfPage,
    atTopOfPage,
    buildSpacerInternalId,
    getDragItemDocumentTop,
    getDragItemId,
    getDragItemIdUnderTarget,
    getDragItemOffsetTop,
    getDragItemWidth,
    handleScroll,
    preventDefault,
    scrollDown,
    scrollUp,
    targetIsDragButton
} from "./backlogItemPlanningPanelUtils";
import { addActionButtons } from "./backlogItemPlanningPanelJsxUtils";
import { productBacklogItemMenuBuilder } from "../../../common/itemMenuBuilders";
import { BELOW_LAST_CARD_ID } from "./backlogItemPlanningPanelConsts";

/* exported components */

export const buildDragBacklogItemElt = (
    editMode: EditMode,
    item: SaveableBacklogItem,
    renderMobile: boolean,
    offsetTop: number,
    width: any
) => {
    const itemEventHandlers: ItemMenuEventHandlers = {
        handleEvent: (eventName: string, itemId: string) => {
            throw Error(`unexpected condition- event ${eventName} should not be triggered while element is being dragged`);
        }
    };
    return (
        <BacklogItemCard
            buildItemMenu={productBacklogItemMenuBuilder(itemEventHandlers)}
            estimate={item.estimate}
            hasDetails={editMode === EditMode.Edit}
            internalId={`${item.id}`}
            isDraggable={editMode === EditMode.Edit}
            isSelectable={editMode === EditMode.Edit}
            itemId={buildBacklogDisplayId(item.externalId, item.friendlyId)}
            itemType={item.type === "story" ? BacklogItemTypeEnum.Story : BacklogItemTypeEnum.Bug}
            key={buildBacklogItemKey(item)}
            marginBelowItem
            offsetTop={offsetTop}
            reasonText={item.reasonPhrase}
            renderMobile={renderMobile}
            roleText={item.rolePhrase}
            showDetailMenu={false}
            status={item.status}
            titleText={item.storyPhrase}
            width={width}
        />
    );
};

export const InnerBacklogItemPlanningPanel: React.FC<BacklogItemPlanningPanelProps> = (props) => {
    logger.info("render(InnerBacklogItemPlanningPanel)", [loggingTags.DRAG_BACKLOGITEM]);
    // #region state with refs

    // 1. Refs that are not used for re-rendering

    const cardPositionsRef = React.useRef();
    const setCardPositions = (data) => {
        cardPositionsRef.current = data;
    };
    const dragItemStartDocumentTopRef = React.useRef<number>();
    const setDragItemStartDocumentTop = (data) => {
        dragItemStartDocumentTopRef.current = data;
    };
    const dragItemStartOffsetTopRef = React.useRef<number>();
    const setDragItemStartOffsetTop = (data) => {
        dragItemStartOffsetTopRef.current = data;
    };
    const dragStartClientYRef = React.useRef<number>();
    const setDragStartClientY = (data) => {
        dragStartClientYRef.current = data;
    };
    const dragStartScrollYRef = React.useRef<number>();
    const setDragStartScrollY = (data) => {
        dragStartScrollYRef.current = data;
    };
    const itemCardWidthRef = React.useRef();
    const setItemCardWidth = (data) => {
        itemCardWidthRef.current = data;
    };

    // 2. Refs that are used for re-rendering

    const [isDragging, _setIsDragging] = useState(false);
    const isDraggingRef = React.useRef();
    const setIsDragging = (data) => {
        logger.info(`setIsDragging:${data}`, [loggingTags.DRAG_BACKLOGITEM]);
        isDraggingRef.current = data;
        _setIsDragging(data);
    };
    const [dragItemId, _setDragItemId] = useState(null);
    const dragItemIdRef = React.useRef();
    const setDragItemId = (data) => {
        dragItemIdRef.current = data;
        _setDragItemId(data);
    };
    const [dragOverItemId, _setDragOverItemId] = useState(null);
    const dragOverItemIdRef = React.useRef();
    const setDragOverItemId = (data) => {
        dragOverItemIdRef.current = data;
        _setDragOverItemId(data);
    };
    const [dragItemDocumentTop, _setDragItemDocumentTop] = useState(null);
    const dragItemDocumentTopRef = React.useRef();
    const setDragItemDocumentTop = (data) => {
        dragItemDocumentTopRef.current = data;
        _setDragItemDocumentTop(data);
    };
    const [dragItemOffsetTop, _setDragItemOffsetTop] = useState(null);
    const dragItemOffsetTopRef = React.useRef<number>();
    const setDragItemOffsetTop = (data) => {
        dragItemOffsetTopRef.current = data;
        _setDragItemOffsetTop(data);
    };
    const [dragItemClientY, _setDragItemClientY] = useState(null);
    const dragItemClientYRef = React.useRef();
    const setDragItemClientY = (data) => {
        dragItemClientYRef.current = data;
        _setDragItemClientY(data);
    };
    // #endregion

    useRecursiveTimeout(() => {
        if (isDraggingRef.current) {
            if (atBottomOfPage(dragItemClientYRef.current)) {
                scrollDown((deltaY) => {
                    handleScroll(
                        deltaY,
                        cardPositionsRef.current,
                        dragItemDocumentTopRef.current,
                        dragItemOffsetTopRef.current,
                        setDragItemDocumentTop,
                        setDragItemOffsetTop,
                        dragOverItemIdRef.current,
                        setDragOverItemId
                    );
                });
            }
            if (atTopOfPage(dragItemClientYRef.current)) {
                scrollUp((deltaY) => {
                    handleScroll(
                        deltaY,
                        cardPositionsRef.current,
                        dragItemDocumentTopRef.current,
                        dragItemOffsetTopRef.current,
                        setDragItemDocumentTop,
                        setDragItemOffsetTop,
                        dragOverItemIdRef.current,
                        setDragOverItemId
                    );
                });
            }
        }
    }, 500);

    const onMouseDown = (e: React.BaseSyntheticEvent<HTMLDivElement>, clientY: number) => {
        const logContainer = logger.info(`onMouseDown:${clientY}`, [loggingTags.DRAG_BACKLOGITEM]);
        if (targetIsDragButton(e)) {
            logger.info("target is drag button", [loggingTags.DRAG_BACKLOGITEM], logContainer);
            const id = getDragItemId(e.target);
            if (id) {
                const width = getDragItemWidth(e.target);
                const top = getDragItemDocumentTop(e.target);
                const offsetTop = getDragItemOffsetTop(e.target);
                setDragItemOffsetTop(offsetTop);
                setDragItemStartOffsetTop(offsetTop);
                setDragItemDocumentTop(top);
                setDragItemStartDocumentTop(top);
                setDragStartClientY(clientY);
                setDragStartScrollY(window.scrollY);
                setItemCardWidth(width);
                setDragItemId(id);
                setDragOverItemId(id);
            }
            preventDefault(e);
            return false;
        }
        logger.info("target is NOT drag button", [loggingTags.DRAG_BACKLOGITEM], logContainer);
        return true;
    };

    const getMouseDragDistance = (clientY: number) => {
        const dragStartClientY = dragStartClientYRef.current;
        const dragStartScrollY = dragStartScrollYRef.current;
        const mouseClientYToDocumentTop = clientY + window.scrollY;
        const mouseStartClientYToDocumentTop = dragStartClientY + dragStartScrollY;
        const mouseDeltaClientY = mouseClientYToDocumentTop - mouseStartClientYToDocumentTop;
        return mouseDeltaClientY;
    };

    enum EventType {
        Mouse,
        Touch
    }

    const onMouseMove = (e: React.BaseSyntheticEvent<HTMLDivElement>, clientY: number, eventType: EventType = EventType.Mouse) => {
        const logContext = logger.info(`onMouseMove:${clientY}`, [loggingTags.DRAG_BACKLOGITEM]);
        if (dragStartClientYRef.current) {
            if (eventType === EventType.Touch) {
                preventDefault(e);
            }
            const mouseDeltaClientY = getMouseDragDistance(clientY);
            logger.info(`mouseDelta:${mouseDeltaClientY}`, [loggingTags.DRAG_BACKLOGITEM], logContext);
            if (Math.abs(mouseDeltaClientY) > 5) {
                setIsDragging(true);
            }
            if (isDraggingRef.current) {
                const top = dragItemStartDocumentTopRef.current + mouseDeltaClientY;
                const offsetTop = dragItemStartOffsetTopRef.current + mouseDeltaClientY;
                setDragItemOffsetTop(offsetTop);
                setDragItemDocumentTop(top);
                setDragItemClientY(clientY);
                const adjustY = dragItemStartDocumentTopRef.current - dragStartScrollYRef.current - dragStartClientYRef.current;
                const overItemId = getDragItemIdUnderTarget(e, clientY, adjustY, cardPositionsRef.current);
                if (overItemId) {
                    setDragOverItemId(overItemId);
                }
            }
            return false;
        }
        logger.info("not dragging, bypassed preventDefault", [loggingTags.DRAG_BACKLOGITEM], logContext);
        return true;
    };

    const isLastCard = (id: string) => id === lastItemId;

    const getPrevIdForItem = (id: string) => {
        let idx = 0;
        let lastId: string = null;
        let result: string = null;
        while (idx < props.allItems.length && !result) {
            const item = props.allItems[idx];
            if (item.id === id) {
                result = lastId;
            }
            lastId = item.id;
            idx++;
        }
        return result;
    };

    const draggingAboveTopItem = (dragOverItemId: string, dragItemId: string) => dragOverItemId === dragItemId;

    const draggingBelowBottomItem = (dragOverItemId: string, dragItemId: string) =>
        dragOverItemId === BELOW_LAST_CARD_ID && isLastCard(dragItemId);

    const draggingOverSameMiddleItem = (dragOverItemId: string, dragItemId: string) =>
        dragItemId === getPrevIdForItem(dragOverItemId);

    const isOverSameCard = (overItemId: string, itemId: string) => {
        return (
            draggingBelowBottomItem(overItemId, itemId) ||
            draggingAboveTopItem(overItemId, itemId) ||
            draggingOverSameMiddleItem(overItemId, itemId)
        );
    };

    const getTrueItemId = (id: string) => (id === BELOW_LAST_CARD_ID ? null : id);

    const onMouseUp = (e: React.BaseSyntheticEvent<HTMLDivElement>) => {
        logger.info(`onMouseUp`, [loggingTags.DRAG_BACKLOGITEM]);

        if (isDraggingRef.current) {
            if (props.onReorderBacklogItems) {
                if (!isOverSameCard(dragOverItemIdRef.current, dragItemIdRef.current)) {
                    props.onReorderBacklogItems(dragItemIdRef.current, getTrueItemId(dragOverItemIdRef.current));
                }
            }
            setIsDragging(false);
            setDragStartClientY(null);
            setDragItemId(null);
            setDragOverItemId(null);
            return false;
        } else if (dragStartClientYRef.current) {
            setDragStartClientY(null);
            setDragItemId(null);
            setDragOverItemId(null);
            return false;
        }
        return true;
    };

    const ref = React.createRef<HTMLDivElement>();
    React.useEffect(() => {
        logger.info("setting up event listeners for touchstart, touchmove, touchend", [loggingTags.DRAG_BACKLOGITEM]);
        const listenerHost = window;
        const removeEventListener = listenerHost.removeEventListener;
        let touchStartListener;
        let touchMoveListener;
        let touchEndListener;
        if (listenerHost) {
            logger.info("listenerHost set", [loggingTags.DRAG_BACKLOGITEM]);
            touchStartListener = listenerHost.addEventListener(
                "touchstart",
                (e: any) => {
                    const logInfo = logger.info("touchstart", [loggingTags.DRAG_BACKLOGITEM]);
                    if (isSingleTouch(e.touches)) {
                        const result = onMouseDown(e, touchesToClientY(e.touches));
                        logger.info(`isSingleTouch, returning ${result}`, [loggingTags.DRAG_BACKLOGITEM], logInfo);
                        return result;
                    }
                    logger.info("not single touch, returning true", [loggingTags.DRAG_BACKLOGITEM], logInfo);
                    return true;
                },
                { passive: false }
            );
            touchMoveListener = listenerHost.addEventListener(
                "touchmove",
                (e: any) => {
                    const result = onMouseMove(e, touchesToClientY(e.touches), EventType.Touch);
                    logger.info(`touchmove, returning ${result}`, [loggingTags.DRAG_BACKLOGITEM]);
                    return result;
                },
                { passive: false }
            );
            touchEndListener = listenerHost.addEventListener(
                "touchend",
                (e: any) => {
                    const result = onMouseUp(e);
                    logger.info(`touchend, returning ${result}`, [loggingTags.DRAG_BACKLOGITEM]);
                    return result;
                },
                { passive: false }
            );
        }
        const cleanup = () => {
            logger.info("cleanup event listeners for touchstart, touchmove, touchend", [loggingTags.DRAG_BACKLOGITEM]);
            removeEventListener("touchstart", touchStartListener);
            removeEventListener("touchmove", touchMoveListener);
            removeEventListener("touchend", touchEndListener);
        };
        return cleanup;
    }, []);
    React.useEffect(() => {
        const newCardPositions: CardPosition[] = [];
        const computedStyle = getComputedStyle(ref.current);
        const width = ref.current?.offsetWidth - parseFloat(computedStyle.paddingLeft) - parseFloat(computedStyle.paddingRight);
        const elts = ref.current.querySelectorAll(`[data-class='backlogitem']`);
        elts.forEach((elt) => {
            const id = elt.getAttribute("data-id");
            const rect = elt.getBoundingClientRect();
            newCardPositions.push({ id, documentTop: rect.top + window.scrollY, documentBottom: rect.bottom + window.scrollY });
        });
        setCardPositions(newCardPositions);
        setItemCardWidth(width);
    }, [props.allItems, props.editMode, props.renderMobile]);
    const dispatch = useDispatch();
    let inLoadedSection = false;
    let inAddedSection = false;
    let afterPushedItem = false;
    let renderElts = [];
    let suppressTopPadding = true;
    let lastItemWasUnsaved = false;
    let lastItemId: string = null;
    props.allItems.forEach((item) => {
        const isDragItem = isDraggingRef.current && dragItemIdRef.current === item.id;
        const isDragOverItem = isDraggingRef.current && dragOverItemIdRef.current === item.id;
        if (isDragItem) {
            const elt = buildDragBacklogItemElt(
                props.editMode,
                item,
                props.renderMobile,
                dragItemOffsetTopRef.current,
                itemCardWidthRef.current
            );
            renderElts.push(elt);
        }
        let highlightAbove = false;
        if (item.source === Source.Added) {
            inAddedSection = true;
            highlightAbove = afterPushedItem;
        }
        if (item.source === Source.Loaded) {
            highlightAbove = afterPushedItem;
            if (inAddedSection) {
                renderElts.push(<SimpleDivider key={buildDividerKey(item)} />);
                inAddedSection = false;
            }
            if (!inLoadedSection) {
                const suppressButtonSpacing = !!props.renderMobile;
                addActionButtons(
                    renderElts,
                    props.editMode,
                    suppressTopPadding || lastItemWasUnsaved,
                    suppressButtonSpacing,
                    props.onAddNewBacklogItemForm
                );
            }
            inLoadedSection = true;
        }
        if (item.source === Source.Added || item.source === Source.Loaded) {
            if (isDragOverItem) {
                const cardKey = `none---${item.id}---none`;
                renderElts.push(<SimpleDivider key={`divider-${cardKey}`} />);
                const showDetailMenu = item.id === props.openedDetailMenuBacklogItemId;
                const itemEventHandlers: ItemMenuEventHandlers = {
                    handleEvent: (eventName: string, itemId: string) => {
                        if (eventName === "onRemoveItemClicked") {
                            dispatch(apiDeleteBacklogItem(item.id));
                        } else if (eventName === "onEditItemClicked") {
                            dispatch(editBacklogItem(item.id));
                        } else {
                            throw Error(`${eventName} is not handled`);
                        }
                    }
                };
                const dragOverItemBacklogItemCard = (
                    <BacklogItemCard
                        key={cardKey}
                        buildItemMenu={productBacklogItemMenuBuilder(itemEventHandlers)}
                        estimate={null}
                        hasDetails={props.editMode === EditMode.Edit}
                        internalId={buildSpacerInternalId(item.id)}
                        isDraggable={props.editMode === EditMode.Edit}
                        isSelectable={props.editMode === EditMode.Edit}
                        itemId={null}
                        itemType={BacklogItemTypeEnum.None}
                        marginBelowItem
                        reasonText={null}
                        renderMobile={props.renderMobile}
                        roleText={null}
                        status={item.status}
                        titleText={null}
                        onDetailClicked={() => {
                            dispatch(backlogItemDetailClicked(item.id));
                        }}
                        onCheckboxChange={(checked) => {
                            if (checked) {
                                dispatch(selectProductBacklogItem(item.id));
                            } else {
                                dispatch(unselectProductBacklogItem(item.id));
                            }
                        }}
                        showDetailMenu={showDetailMenu}
                    />
                );
                renderElts.push(dragOverItemBacklogItemCard);
            }
            const showItem = !isDragItem;
            const showDetailMenu = item.id === props.openedDetailMenuBacklogItemId;
            renderElts.push(
                <BacklogItemPlanningItem
                    key={buildBacklogItemPlanningItemKey(item)}
                    {...item}
                    editMode={props.editMode}
                    renderMobile={props.renderMobile}
                    highlightAbove={highlightAbove}
                    suppressTopPadding={suppressTopPadding || lastItemWasUnsaved}
                    showDetailMenu={showDetailMenu}
                    hidden={!showItem}
                />
            );
        }
        afterPushedItem = item.source === Source.Pushed;
        lastItemWasUnsaved = item.source === Source.Added && !item.saved;
        suppressTopPadding = false;
        lastItemId = item.id;
    });
    if (!inLoadedSection) {
        const suppressButtonSpacing = !!props.renderMobile;
        addActionButtons(
            renderElts,
            props.editMode,
            suppressTopPadding || lastItemWasUnsaved,
            suppressButtonSpacing,
            props.onAddNewBacklogItemForm
        );
    }
    if (inLoadedSection) {
        renderElts.push(<SimpleDivider key="last-divider" />);
    }

    const classNameToUse = buildClassName(css.backlogItemPlanningPanel, props.className);
    /* NOTE: Handlers were here */
    const isSingleTouch = (touches: React.TouchList) => touches.length === 1;
    const touchesToClientY = (touches: React.TouchList): number | null => {
        if (touches.length === 1) {
            return touches[0].clientY;
        }
        return null;
    };
    return (
        <div
            ref={ref}
            className={classNameToUse}
            onMouseDown={(e: React.MouseEvent<HTMLDivElement>) => onMouseDown(e as any, e.clientY)}
            onMouseMove={(e: React.MouseEvent<HTMLDivElement>) => onMouseMove(e as any, e.clientY)}
            onMouseUp={(e: React.MouseEvent<HTMLDivElement>) => onMouseUp(e as any)}
        >
            {renderElts}
        </div>
    );
};

export const BacklogItemPlanningPanel = withTranslation()(InnerBacklogItemPlanningPanel);
