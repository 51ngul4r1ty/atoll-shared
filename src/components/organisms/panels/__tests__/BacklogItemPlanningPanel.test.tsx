// externals
import * as React from "react";

// test related
import "jest";
import { render } from "@testing-library/react";

// code under test
import { BacklogItemPlanningPanel } from "../BacklogItemPlanningPanel";
import { EditMode } from "../../../molecules/buttons/EditButton";
import { BacklogItemWithSource, BacklogItemSource } from "../../../../reducers/backlogItemsReducer";

// mocks
const mockUseDispatch = jest.fn();
jest.mock("react-redux", () => ({
    useDispatch: () => mockUseDispatch()
}));
jest.mock("../../forms/BacklogItemDetailForm", () => ({
    BacklogItemDetailForm: (obj) => {
        return (
            `[Mock BacklogItemDetailForm className="${obj.className}" instanceId=${obj.instanceId} ` +
            `externalId=${obj.externalId} id=${obj.id}]`
        );
    }
}));
jest.mock("../../../molecules/cards/BacklogItemCard", () => {
    const result = {
        ...jest.requireActual("../../../molecules/cards/BacklogItemCard"),
        BacklogItemCard: (obj) =>
            `[Mock BacklogItemCard internalId=${obj.internalId} itemId=${obj.itemId} ` +
            `titleText="${obj.titleText}" marginBelowItem=${obj.marginBelowItem}]`
    };
    return result;
});
jest.mock("../../../atoms/dividers/SimpleDivider", () => ({
    SimpleDivider: () => "[Mock SimpleDivider]"
}));
jest.mock("../../../molecules/buttons/AddButton", () => ({
    AddButton: (obj) => `[Mock AddButton itemName=${obj.itemName}]`
}));
jest.mock("react-i18next", () => ({
    withTranslation: () => (elt) => elt
}));

const sourceToString = (source: BacklogItemSource) => {
    let result: string;
    switch (source) {
        case BacklogItemSource.Added: {
            result = "added";
            break;
        }
        case BacklogItemSource.Loaded: {
            result = "loaded";
            break;
        }
        case BacklogItemSource.Pushed: {
            result = "pushed";
            break;
        }
        default: {
            result = `unknown:${source}`;
        }
    }
    return result;
};

const buildExternalId = (source: BacklogItemSource, itemNumber: number) => {
    const suffix = sourceToString(source);
    return `item-${itemNumber}-${suffix}`;
};

const buildStoryPhrase = (source: BacklogItemSource, itemNumber: number) => {
    const sourceText = sourceToString(source);
    return `${sourceText} story ${itemNumber}`;
};

const buildId = (itemNumber: number) => {
    return `fake-id-${itemNumber}`;
};

const buildCommonItem = (
    source: BacklogItemSource,
    itemNumber: number,
    instanceId: number | null,
    estimate: number,
    saved: boolean = true
): BacklogItemWithSource => ({
    createdAt: new Date(),
    estimate,
    externalId: buildExternalId(source, itemNumber),
    id: buildId(itemNumber),
    instanceId,
    reasonPhrase: null,
    rolePhrase: null,
    saved,
    source,
    storyPhrase: buildStoryPhrase(source, itemNumber),
    type: "story"
});

const buildAddedItem = (itemNumber: number, instanceId: number | null, estimate: number, saved: boolean): BacklogItemWithSource =>
    buildCommonItem(BacklogItemSource.Added, itemNumber, instanceId, estimate, saved);

const buildLoadedItem = (itemNumber: number, estimate: number): BacklogItemWithSource =>
    buildCommonItem(BacklogItemSource.Loaded, itemNumber, null, estimate);

describe("BacklogItemPlanningPanel", () => {
    it("should include spacing between added items, action buttons and loaded items", () => {
        const allItems = [];
        allItems.push(buildAddedItem(1, 1000, 13, false));
        allItems.push(buildLoadedItem(1, 8));
        allItems.push(buildLoadedItem(2, 3));
        const editMode = EditMode.Edit;
        const wrapper = render(
            <BacklogItemPlanningPanel
                openedDetailMenuBacklogItemId={null}
                allItems={allItems}
                editMode={editMode}
                renderMobile
                onAddNewBacklogItem={() => {}}
                onReorderBacklogItems={() => {}}
            />
        );
        expect(wrapper.container).toMatchSnapshot();
    });
    it("should include spacing between added saved, added unsaved, and loaded item", () => {
        const allItems = [];
        allItems.push(buildAddedItem(1, 2000, 13, true));
        allItems.push(buildAddedItem(2, 2001, 5, false));
        allItems.push(buildLoadedItem(3, 8));
        const editMode = EditMode.Edit;
        const wrapper = render(
            <BacklogItemPlanningPanel
                openedDetailMenuBacklogItemId={null}
                allItems={allItems}
                editMode={editMode}
                renderMobile
                onAddNewBacklogItem={() => {}}
                onReorderBacklogItems={() => {}}
            />
        );
        expect(wrapper.container).toMatchSnapshot();
    });
    it("should not include double spacing between added saved, added unsaved and action buttons", () => {
        const allItems = [];
        allItems.push(buildAddedItem(1, 3000, 13, true));
        allItems.push(buildAddedItem(2, 3001, 5, false));
        const editMode = EditMode.Edit;
        const wrapper = render(
            <BacklogItemPlanningPanel
                openedDetailMenuBacklogItemId={null}
                allItems={allItems}
                editMode={editMode}
                renderMobile
                onAddNewBacklogItem={() => {}}
                onReorderBacklogItems={() => {}}
            />
        );
        expect(wrapper.container).toMatchSnapshot();
    });
    it("should handle scenario where unsaved form visible and user switches from edit to view mode", () => {
        const allItems = [];
        allItems.push(buildAddedItem(1, 4000, 13, false));
        allItems.push(buildLoadedItem(3, 8));
        const editMode = EditMode.View;
        const wrapper = render(
            <BacklogItemPlanningPanel
                openedDetailMenuBacklogItemId={null}
                allItems={allItems}
                editMode={editMode}
                renderMobile
                onAddNewBacklogItem={() => {}}
                onReorderBacklogItems={() => {}}
            />
        );
        expect(wrapper.container).toMatchSnapshot();
    });
    it("should handle scenario where 3 unsaved added items shown in sequence", () => {
        const allItems = [];
        allItems.push(buildAddedItem(1, 5000, 13, false));
        allItems.push(buildAddedItem(1, 5001, 8, false));
        allItems.push(buildAddedItem(1, 5002, 5, false));
        allItems.push(buildLoadedItem(3, 8));
        const editMode = EditMode.Edit;
        const wrapper = render(
            <BacklogItemPlanningPanel
                openedDetailMenuBacklogItemId={null}
                allItems={allItems}
                editMode={editMode}
                renderMobile
                onAddNewBacklogItem={() => {}}
                onReorderBacklogItems={() => {}}
            />
        );
        expect(wrapper.container).toMatchSnapshot();
    });
});
