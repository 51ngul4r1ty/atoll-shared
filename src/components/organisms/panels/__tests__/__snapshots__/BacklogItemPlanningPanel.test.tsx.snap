// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BacklogItemPlanningPanel should handle scenario where 3 unsaved added items shown in sequence 1`] = `
<div>
  <div
    class="backlogItemPlanningPanel mobile"
  >
    [Mock SimpleDivider]
    [Mock BacklogItemDetailForm className="backlogItemUserStoryFormRow" instanceId=5000 externalId=item-1-added id=fake-id-1]
    [Mock SimpleDivider]
    [Mock BacklogItemDetailForm className="backlogItemUserStoryFormRow" instanceId=5001 externalId=item-2-added id=fake-id-2]
    [Mock SimpleDivider]
    [Mock BacklogItemDetailForm className="backlogItemUserStoryFormRow" instanceId=5002 externalId=item-3-added id=fake-id-3]
    [Mock SimpleDivider]
    <div
      class="backlogItemPlanningActionPanel mobile"
    >
      [Mock AddButton itemName=story]
      [Mock AddButton itemName=issue]
    </div>
    [Mock SimpleDivider]
    [Mock BacklogItemCard internalId=fake-id-4 itemId=item-4-loaded titleText="loaded story 4" marginBelowItem=true]
    [Mock SimpleDivider]
  </div>
</div>
`;

exports[`BacklogItemPlanningPanel should handle scenario where unsaved form visible and user switches from edit to view mode 1`] = `
<div>
  <div
    class="backlogItemPlanningPanel mobile"
  >
    [Mock SimpleDivider]
    [Mock BacklogItemCard internalId=fake-id-1 itemId=item-1-added titleText="added story 1" marginBelowItem=true]
    [Mock SimpleDivider]
    [Mock SimpleDivider]
    [Mock BacklogItemCard internalId=fake-id-2 itemId=item-2-loaded titleText="loaded story 2" marginBelowItem=true]
    [Mock SimpleDivider]
  </div>
</div>
`;

exports[`BacklogItemPlanningPanel should include spacing between added items, action buttons and loaded items 1`] = `
<div>
  <div
    class="backlogItemPlanningPanel mobile"
  >
    [Mock SimpleDivider]
    [Mock BacklogItemDetailForm className="backlogItemUserStoryFormRow" instanceId=1000 externalId=item-1-added id=fake-id-1]
    [Mock SimpleDivider]
    <div
      class="backlogItemPlanningActionPanel mobile"
    >
      [Mock AddButton itemName=story]
      [Mock AddButton itemName=issue]
    </div>
    [Mock SimpleDivider]
    [Mock BacklogItemCard internalId=fake-id-2 itemId=item-2-loaded titleText="loaded story 2" marginBelowItem=true]
    [Mock SimpleDivider]
    [Mock BacklogItemCard internalId=fake-id-3 itemId=item-3-loaded titleText="loaded story 3" marginBelowItem=true]
    [Mock SimpleDivider]
  </div>
</div>
`;

exports[`BacklogItemPlanningPanel should include spacing between added saved, added unsaved, and loaded item 1`] = `
<div>
  <div
    class="backlogItemPlanningPanel mobile"
  >
    [Mock SimpleDivider]
    [Mock BacklogItemCard internalId=fake-id-1 itemId=item-1-added titleText="added story 1" marginBelowItem=true]
    [Mock SimpleDivider]
    [Mock BacklogItemDetailForm className="backlogItemUserStoryFormRow embeddedBacklogItemUserStoryFormRow" instanceId=2001 externalId=item-2-added id=fake-id-2]
    [Mock SimpleDivider]
    <div
      class="backlogItemPlanningActionPanel mobile"
    >
      [Mock AddButton itemName=story]
      [Mock AddButton itemName=issue]
    </div>
    [Mock SimpleDivider]
    [Mock BacklogItemCard internalId=fake-id-3 itemId=item-3-loaded titleText="loaded story 3" marginBelowItem=true]
    [Mock SimpleDivider]
  </div>
</div>
`;

exports[`BacklogItemPlanningPanel should not include double spacing between added saved, added unsaved and action buttons 1`] = `
<div>
  <div
    class="backlogItemPlanningPanel mobile"
  >
    [Mock SimpleDivider]
    [Mock BacklogItemCard internalId=fake-id-1 itemId=item-1-added titleText="added story 1" marginBelowItem=true]
    [Mock SimpleDivider]
    [Mock BacklogItemDetailForm className="backlogItemUserStoryFormRow embeddedBacklogItemUserStoryFormRow" instanceId=3001 externalId=item-2-added id=fake-id-2]
    <div
      class="backlogItemPlanningActionPanel mobile"
    >
      [Mock AddButton itemName=story]
      [Mock AddButton itemName=issue]
    </div>
  </div>
</div>
`;
